credentials/google/google.go

```diff
@@ -64,14 +64,25 @@ func NewDefaultCredentials() credentials.Bundle {
 //
 // This API is experimental.
 func NewComputeEngineCredentials() credentials.Bundle {
+	return NewComputeEngineCredsWithPerRPC(nil)
+}
+
+// NewComputeEngineCredsWithPerRPC returns a credentials bundle that is configured to work
+// with google services. This API must only be used when running on GCE.
+//
+// This API is experimental.
+func NewComputeEngineCredsWithPerRPC(perRPC credentials.PerRPCCredentials) credentials.Bundle {
 	c := &creds{
 		newPerRPCCreds: func() credentials.PerRPCCredentials {
+			if perRPC != nil {
+				return perRPC
+			}
 			return oauth.NewComputeEngine()
 		},
 	}
```

Comment:
Let's make this a little more efficient:

```go
c := &creds{newPerRPCCreds: perRPC}
if c.newPerRPCCreds == nil {
	c.newPerRPCCreds = oauth.NewComputeEngine
}
```

Reply:
Hmm, wait, this doesn't work, exactly, because of the func returning `PerRPCCredentials` vs. _actual_ `PerRPCCredentials`.  Why do we have a factory instead of the credentials, here?

Reply:
The creds struct has both `perRPCCreds` and `newPerRPCCreds`.
The factory function is only called in `NewWithMode`.

It's so that no per-RPC is shared by two bundles?
But then why is it OK to share transport creds?

(Just discussing, not suggesting changes to this PR)

Reply:
Oh, actually, transport creds are also not shared. `NewWithMode()` also creates them.

Reply:
> It's so that no per-RPC is shared by two bundles?
> But then why is it OK to share transport creds?

`TransportCredentials` have a `Clone` method, but `PerRPCCredentials` do not.

But why is it not okay to reuse `PerRPCCredentials`?  There are no concerns about initialization or concurrency AFAICT.

Reply:
We can clean that up later.  The comment here would be to avoid checking a conditional _inside_ the closure; instead check outside of the closure and create the `&creds` differently from the start.